File: platform/desktop-app/window_recall.py
"""Local-only window recall control; never sends keyboard events to a peer."""
import json
import os
from pathlib import Path
import re
import subprocess
import time

DEFAULT_SHORTCUT = 'Ctrl+Alt+Shift+H'


def canonical_shortcut(value):
    names = {'CTRL': 'Ctrl', 'CONTROL': 'Ctrl', 'ALT': 'Alt', 'OPTION': 'Alt',
             'SHIFT': 'Shift', 'SUPER': 'Super', 'WIN': 'Super', 'CMD': 'Super', 'COMMAND': 'Super'}
    modifiers = set()
    letter = None
    for token in value.upper().split('+'):
        token = token.strip()
        if token in names:
            name = names[token]
            if name in modifiers: raise ValueError('快捷键修饰键重复')
            modifiers.add(name)
        elif re.fullmatch('[A-Z]', token) and letter is None: letter = token
        else: raise ValueError('请输入修饰键和一个字母，例如 Ctrl+Alt+Shift+H')
    if not modifiers or letter is None: raise ValueError('快捷键需要修饰键和一个字母')
    return '+'.join([name for name in ('Ctrl', 'Alt', 'Shift', 'Super') if name in modifiers] + [letter])


class RecallController:
    def __init__(self, platform, binary, data, shortcut=DEFAULT_SHORTCUT):
        self.platform, self.binary, self.data = platform, binary, Path(data)
        try: self.shortcut = canonical_shortcut(shortcut)
        except (ValueError, AttributeError): self.shortcut = DEFAULT_SHORTCUT
        self.process = None
        self.log = None
        self.status = ''
        self.retry_at = 0
        self.previous = None

    def _run(self, args):
        options = {'creationflags': subprocess.CREATE_NO_WINDOW} if self.platform == 'windows' else {}
        result = subprocess.run(args, capture_output=True, text=True, errors='replace', timeout=4, **options)
        if result.returncode: raise RuntimeError(result.stderr.strip() or result.stdout.strip() or '收回窗口组件执行失败')
        return result.stdout.strip()

    def _linux(self, expression):
        output = self._run(['hyprctl', 'eval', expression])
        # Hyprctl may return a compositor error with a zero process exit code.
        if 'error' in output.lower() or 'nil value' in output.lower(): raise RuntimeError(output)
        return output

    def configure(self, value):
        value = canonical_shortcut(value)
        if self.platform == 'linux':
            self._linux('return viewflow.set_recall_shortcut(' + json.dumps(value) + ')')
            self.shortcut = value
            self.status = '收回本机窗口：' + value
            self.retry_at = time.monotonic() + 5
        else:
            self._run([str(self.binary), '--validate-shortcut', '--shortcut', value])
            self.previous = self.shortcut if self.process else None
            self.stop()
            self.shortcut = value
            self._start()

    def _start(self):
        self.data.mkdir(parents=True, exist_ok=True)
        self.log = (self.data / 'window-recall.log').open('wb')
        try:
            self.process = subprocess.Popen([str(self.binary), '--watch', '--shortcut', self.shortcut, '--parent', str(os.getpid())],
                                            stdin=subprocess.DEVNULL, stdout=self.log, stderr=self.log,
                                            creationflags=subprocess.CREATE_NO_WINDOW)
            self.status = '正在注册收回窗口快捷键'
        except Exception:
            self.log.close(); self.log = None
            raise

    def poll(self):
        try:
            if self.platform == 'linux':
                if time.monotonic() >= self.retry_at:
                    self.retry_at = time.monotonic() + 5
                    self.configure(self.shortcut)
            elif self.process is None and time.monotonic() >= self.retry_at:
                self._start()
            elif self.process is not None:
                if self.process.poll() is not None:
                    message = (self.data / 'window-recall.log').read_text(errors='replace').strip()
                    previous = self.previous
                    self.stop(); self.previous = None; self.retry_at = time.monotonic() + 5
                    if previous and previous != self.shortcut:
                        self.shortcut = previous
                        self._start()
                    raise RuntimeError(message or '收回窗口快捷键未注册')
                elif 'recall ready' in (self.data / 'window-recall.log').read_text(errors='replace'):
                    self.previous = None
                    self.status = '收回本机窗口：' + self.shortcut
        except (OSError, ValueError, RuntimeError, subprocess.SubprocessError) as error:
            self.status = str(error)
            self.retry_at = time.monotonic() + 5
        return self.status

    def recall(self):
        if self.platform == 'linux':
            return self._linux('return viewflow.recall_windows(false)')
        return self._run([str(self.binary), '--once'])

    def stop(self):
        if self.process:
            self.process.terminate()
            try: self.process.wait(timeout=2)
            except subprocess.TimeoutExpired: self.process.kill(); self.process.wait(timeout=2)
            self.process = None
        if self.log: self.log.close(); self.log = None

File: platform/macos-app/Sources/WindowRecall.swift
import AppKit
import Combine
import Darwin

@MainActor final class WindowRecallController: ObservableObject {
    @Published private(set) var shortcut = UserDefaults.standard.string(forKey: "recallShortcut") ?? "Ctrl+Alt+Shift+H"
    @Published private(set) var status = "正在注册收回窗口快捷键"
    private var process: Process?
    private var log: FileHandle?
    private var desired = true
    private var retryAt = Date.distantPast
    private var previous: String?
    private var restarting = false
    private let logURL = ProfileStore.root.appendingPathComponent("run/window-recall.log")
    var isStopped: Bool { process?.isRunning != true }

    static func canonical(_ input: String) throws -> String {
        let aliases = ["CTRL":"Ctrl", "CONTROL":"Ctrl", "ALT":"Alt", "OPTION":"Alt", "SHIFT":"Shift", "SUPER":"Super", "WIN":"Super", "CMD":"Super", "COMMAND":"Super"]
        var modifiers = Set<String>(), letter: String?
        for part in input.uppercased().split(separator: "+", omittingEmptySubsequences: false) {
            let token = part.trimmingCharacters(in: .whitespaces)
            if let name = aliases[token] {
                guard modifiers.insert(name).inserted else { throw ViewflowError.invalid("快捷键修饰键重复") }
            } else if token.utf8.count == 1, let code = token.utf8.first, code >= 65 && code <= 90, letter == nil {
                letter = token
            } else { throw ViewflowError.invalid("请输入修饰键和一个字母，例如 Ctrl+Alt+Shift+H") }
        }
        guard let letter, !modifiers.isEmpty else { throw ViewflowError.invalid("快捷键需要修饰键和一个字母") }
        return (["Ctrl", "Alt", "Shift", "Super"].filter { modifiers.contains($0) } + [letter]).joined(separator: "+")
    }
    func configure(_ input: String) {
        do {
            let value = try Self.canonical(input)
            previous = shortcut; shortcut = value; desired = true; retryAt = .distantPast
            if process?.isRunning == true { restarting = true; process?.terminate() }
            poll()
        } catch { status = error.localizedDescription }
    }
    func poll() {
        if let child = process, !child.isRunning {
            let message = (try? String(contentsOf: logURL, encoding: .utf8)) ?? ""
            process = nil; try? log?.close(); log = nil
            let wasRestarting = restarting; restarting = false
            if desired, !wasRestarting, !message.contains("recall ready") {
                status = message.isEmpty ? "收回窗口快捷键未注册" : String(message.suffix(400))
                if let old = previous { shortcut = old; previous = nil }
                retryAt = Date().addingTimeInterval(5)
                return
            }
        }
        guard desired else { return }
        if process == nil, Date() >= retryAt {
            do {
                let child = Process()
                child.executableURL = try BundleTools.executable("viewflow-window-recall")
                child.arguments = ["--watch", "--shortcut", shortcut, "--parent", String(ProcessInfo.processInfo.processIdentifier)]
                try ProfileStore.writePrivate(Data(), to: logURL)
                log = try FileHandle(forWritingTo: logURL)
                child.standardInput = FileHandle.nullDevice; child.standardOutput = log; child.standardError = log
                try child.run(); process = child; status = "正在注册收回窗口快捷键"
            } catch { status = error.localizedDescription; retryAt = Date().addingTimeInterval(5); try? log?.close(); log = nil }
        } else if !restarting, process?.isRunning == true,
                  let output = try? String(contentsOf: logURL, encoding: .utf8), output.contains("recall ready") {
            previous = nil
            UserDefaults.standard.set(shortcut, forKey: "recallShortcut")
            status = "收回本机窗口：\(shortcut)（Alt 对应 Option）"
        }
    }
    func perform() {
        do {
            let executable = try BundleTools.executable("viewflow-window-recall")
            // The resident watcher retains the last physical-screen selection.
            // --once requests its local action and never waits for a remote peer.
            DispatchQueue.global(qos: .userInitiated).async { [weak self] in
                let child = Process(); child.executableURL = executable; child.arguments = ["--once"]
                let pipe = Pipe(); child.standardOutput = pipe; child.standardError = pipe
                do {
                    try child.run()
                    let data = pipe.fileHandleForReading.readDataToEndOfFile(); child.waitUntilExit()
                    let text = String(data: data, encoding: .utf8) ?? ""
                    DispatchQueue.main.async { self?.status = text.trimmingCharacters(in: .whitespacesAndNewlines) }
                } catch { DispatchQueue.main.async { self?.status = error.localizedDescription } }
            }
        } catch { status = error.localizedDescription }
    }
    func stop() {
        desired = false
        if process?.isRunning == true { process?.terminate() }
    }
    func pollTermination(elapsed: TimeInterval) {
        poll()
        if elapsed >= 2, let process, process.isRunning { kill(process.processIdentifier, SIGKILL) }
    }
}
